File: web/visualizer/annotation-overlay.js
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael

/**
 * The annotation family, drawn from the acceptance report's closure.
 * NL-IR closure ledger row G2.
 *
 * ## This file derives nothing, and that is its entire specification
 *
 * A requirement's evidence is NOT the elements carrying its `sce:req`.
 * It is the dependency closure of those elements: the `<send>` that arms
 * the timer, the `<cancel>` naming that send, the target state's entry
 * actions, the source state's exit actions. Measured over the real
 * fixture, 21 of 65 (requirement, dependency) pairs lie on nodes that
 * carry no id at all — every arming send and every cancel among them.
 *
 * So a walk here that picked elements by `sce:req` would draw a picture
 * disagreeing with the table the machine measures, and NOBODY WOULD FIND
 * OUT: a rendered diagram is not bytes a test can diff. That is the
 * reason the Requirement-closure RFC puts this after the report rather
 * than before it.
 *
 * The overlay therefore ARRIVES, computed by `annotation_overlay()` in
 * the codegen WASM, which calls the report's own `fragment()`. Every
 * function below is a lookup over what arrived. There is no traversal of
 * the SCXML in this file, and there must never be one.
 *
 * ## Matching, without parsing a path
 *
 * SCE names a node `states.armed.transitions[0]`. The graph here is keyed
 * by state id and edge index. Parsing that string in JavaScript would put
 * a second copy of the path grammar in this file, free to fall behind the
 * Rust that writes it — so the producer publishes `state_id` and
 * `transition_index` beside `node_path`, and this file matches on those.
 */

const SCE_ANNOTATION_UNCLAIMED = 'sce-unclaimed';
const SCE_ANNOTATION_CLAIMED = 'sce-claimed';

class AnnotationOverlay {
    /**
     * @param {object} overlay - the parsed `annotation_overlay()` JSON.
     *   Passed in rather than fetched, so the caller owns the transport
     *   and this class stays testable outside a browser.
     */
    constructor(overlay) {
        if (!overlay || typeof overlay !== 'object') {
            throw new Error('AnnotationOverlay: no overlay given');
        }
        if (overlay.v !== 1) {
            // A deployed page outlives the build that produced its data,
            // so an unknown shape is reported rather than half-read.
            throw new Error(`AnnotationOverlay: unknown overlay version ${overlay.v}`);
        }
        this.nodes = overlay.nodes || [];
        this.fragments = overlay.fragments || [];

        // node_path -> node, so a fragment's dependency resolves to the
        // element it names without re-deriving anything.
        this.byPath = new Map();
        for (const node of this.nodes) {
            this.byPath.set(node.node_path, node);
        }
    }

    /** Requirement ids a state claims, or [] when it claims none. */
    requirementsForState(stateId) {
        const node = this.nodes.find(
            (n) => n.node_type === 'state' && n.state_id === stateId
        );
        return node ? node.requirements : [];
    }

    /** Requirement ids a transition claims, by the key the graph uses. */
    requirementsForTransition(stateId, index) {
        const node = this.nodes.find(
            (n) => n.node_type === 'transition'
                && n.state_id === stateId
                && n.transition_index === index
        );
        return node ? node.requirements : [];
    }

    /**
     * Every node path the requirement's evidence depends on.
     *
     * Straight off the wire. The set deliberately includes paths whose
     * node claims nothing — see the header.
     */
    fragmentPaths(requirement) {
        const fragment = this.fragments.find((f) => f.requirement === requirement);
        return fragment ? fragment.dependencies.map((d) => d.node_path) : [];
    }

    /**
     * Whether this node is in the requirement's fragment WITHOUT carrying
     * its id — the case a by-id walk gets wrong, exposed as its own
     * question so a caller can style it and a test can assert it.
     */
    isSupportingNode(nodePath, requirement) {
        const node = this.byPath.get(nodePath);
        if (!node) {
            return false;
        }
        return (
            this.fragmentPaths(requirement).includes(nodePath)
            && !node.requirements.includes(requirement)
        );
    }

    /** Nodes no requirement claims — what the reviewer is looking for. */
    unclaimedPaths() {
        return this.nodes.filter((n) => n.requirements.length === 0).map((n) => n.node_path);
    }

    /**
     * Stamp the graph the visualizer built with what arrived.
     *
     * Mutates in place, adding `requirements` and `annotationClass` to
     * each state and transition. The graph's shape is untouched: this
     * says what each element CLAIMS, never what exists.
     */
    applyTo(states, transitions) {
        for (const state of states || []) {
            state.requirements = this.requirementsForState(state.id);
            state.annotationClass = state.requirements.length
                ? SCE_ANNOTATION_CLAIMED
                : SCE_ANNOTATION_UNCLAIMED;
        }

        // Transitions are keyed by (source state, position in that state),
        // which is what the producer published beside the path.
        const seen = new Map();
        for (const transition of transitions || []) {
            const source = transition.source ?? transition.from;
            const index = seen.get(source) ?? 0;
            seen.set(source, index + 1);
            transition.requirements = this.requirementsForTransition(source, index);
            transition.annotationClass = transition.requirements.length
                ? SCE_ANNOTATION_CLAIMED
                : SCE_ANNOTATION_UNCLAIMED;
        }
        return { states, transitions };
    }
}

/**
 * Ask the codegen WASM for the overlay.
 *
 * The transport, and the only place this file touches the module. The
 * closure is built on the Rust side; nothing here recomputes it.
 */
async function loadAnnotationOverlay(wasmModule, scxmlContent, scxmlName) {
    const json = wasmModule.annotation_overlay(scxmlContent, scxmlName || 'untitled');
    return new AnnotationOverlay(JSON.parse(json));
}

if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        AnnotationOverlay,
        loadAnnotationOverlay,
        SCE_ANNOTATION_CLAIMED,
        SCE_ANNOTATION_UNCLAIMED,
    };
}

File: sce-build/src/annotation_overlay.rs
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
//
//! What a diagram must be told in order to draw the annotation family —
//! NL→IR closure ledger row G2.
//!
//! The visualizer renders no annotation today, so the picture a reviewer
//! reads cannot say which element a requirement claims and which is
//! unclaimed. It cannot be fixed inside the browser, and the reason is
//! recorded rather than assumed (Requirement-closure RFC, Atomic C, Q1):
//!
//! > **E cannot come first.** The visualizer builds its graph in the
//! > browser from its own `DOMParser` walk and lays it out with a
//! > constraint solver; `sce-build-wasm` exports only codegen. Its output
//! > is not bytes a test can compare, and it is a second walk. So E later
//! > draws the SAME closure C derives, not the other way round.
//!
//! ## ⛔ The one thing this module exists to prevent
//!
//! A requirement's evidence is NOT "the nodes carrying its id". It is the
//! dependency closure of those nodes — the arming `<send>`, the `<cancel>`
//! naming it, the target's entry actions, the source's exit actions.
//! Measured over the real pair, **39 of 72 (requirement, dependency)
//! pairs lie on nodes that do not carry the id** ([`crate::acceptance_report`]).
//!
//! A browser that picked nodes by `sce:req` would therefore draw a
//! *different* answer from the one the acceptance report derives, and
//! nobody would find out: a rendered diagram is not bytes a test can
//! diff. So the closure is computed HERE, by calling
//! [`crate::acceptance_report::fragment`] — the report's own function,
//! not a copy of it — and shipped to the browser as data.
//!
//! ⭐ That is why this module has no walk and no traversal of its own. It
//! composes two existing answers: [`crate::requirements_report::walk_nodes`]
//! for what nodes exist and what each claims, and `fragment` for what each
//! requirement depends on. If either moves, this moves with it, because
//! there is nothing here to move independently.
//!
//! ## Why the coordinates travel beside the path
//!
//! `node_path` is SCE's identity for a node (`states.armed.transitions[0]`).
//! The browser's graph is keyed by state id, so something has to bridge
//! the two — and if the browser bridges it by PARSING the path, the path
//! grammar becomes a second implementation living in JavaScript, free to
//! fall behind the one that writes it.
//!
//! So the bridge is published from the side that owns the identity: every
//! node carries the coordinates the browser already has ([`OverlayNode::state_id`],
//! [`OverlayNode::transition_index`]) next to the path SCE knows it by.
//! The browser matches on those and never parses a path.

use serde::Serialize;

use crate::acceptance_report::{fragment, Relation};
use crate::model::SCXMLModel;
use crate::requirements_report::{walk_nodes, ActionSite, NodeSubject};

/// Wire version of the overlay envelope.
///
/// The browser and the CLI read the same producer, so this exists to tell
/// a CACHED page that it is looking at an overlay shape it does not know —
/// a deployed visualizer outlives the build that produced its data.
pub const OVERLAY_WIRE_VERSION: u32 = 1;

/// Why a node is in a requirement's fragment, as the wire spells it.
///
/// A projection of [`Relation`], not a second enumeration: the match
/// below is exhaustive, so a relation added to the report cannot reach
/// the browser as a silent omission — it stops the build here instead.
fn relation_slug(relation: Relation) -> &'static str {
    match relation {
        Relation::Cited => "cited",
        Relation::ArmingSend => "arming-send",
        Relation::CancelOfArmingSend => "cancel-of-arming-send",
        Relation::TargetEntry => "target-entry",
        Relation::SourceExit => "source-exit",
        Relation::TransitionAction => "transition-action",
    }
}

/// One node of the diagram, and what it claims.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct OverlayNode {
    /// SCE's identity for the node.
    pub node_path: String,
    /// `state` / `transition` / `action` / `invoke` — the short tag the
    /// requirements report routes on.
    pub node_type: &'static str,
    /// The state this node is, or sits in. The browser's graph is keyed
    /// by this, so it is what a lookup matches on.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state_id: Option<String>,
    /// Position of the transition within its state, where the node is a
    /// transition or sits in one. With `state_id` this is the browser's
    /// whole key for an edge.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transition_index: Option<usize>,
    /// Requirement ids the node claims, verbatim and in document order.
    ///
    /// ⚠ Empty is the POINT of this field, not a missing value: an
    /// element no requirement claims is what the reviewer is looking for,
    /// and the row is emitted so the diagram can mark it rather than
    /// leave it indistinguishable from an element nobody drew.
    pub requirements: Vec<String>,
}

impl OverlayNode {
    /// Whether no requirement claims this node.
    pub fn is_unclaimed(&self) -> bool {
        self.requirements.is_empty()
    }
}

/// One node a requirement's evidence depends on.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct OverlayDependency {
    pub node_path: String,
    /// Why it is in the fragment — see [`relation_slug`].
    pub relation: &'static str,
}

/// One requirement, and the closure its evidence depends on.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct OverlayFragment {
    /// The id, verbatim. SCE never normalises one.
    pub requirement: String,
    /// The dependency closure, exactly as [`crate::acceptance_report::fragment`]
    /// derives it.
    ///
    /// ⚠ Contains nodes that do NOT carry the id — that is the whole
    /// reason this ships instead of being recomputed in the browser.
    pub dependencies: Vec<OverlayDependency>,
}

/// Everything a diagram needs to draw the annotation family.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct AnnotationOverlay {
    /// Always [`OVERLAY_WIRE_VERSION`].
    pub v: u32,
    /// Every node the shared walk yields, claimed or not.
    pub nodes: Vec<OverlayNode>,
    /// One per requirement the document claims, in sorted order.
    pub fragments: Vec<OverlayFragment>,
}

impl AnnotationOverlay {
    /// Requirement ids the document claims anywhere.
    pub fn claimed_requirements(&self) -> Vec<&str> {
        self.fragments
            .iter()
            .map(|f| f.requirement.as_str())
            .collect()
    }

    /// Nodes no requirement claims.
    pub fn unclaimed(&self) -> impl Iterator<Item = &OverlayNode> {
        self.nodes.iter().filter(|n| n.is_unclaimed())
    }
}

/// Where a node sits, in the coordinates the browser's graph already has.
///
/// Derived from the walk's own [`NodeSubject`], so this is a projection of
/// the shared walk rather than a second opinion about where a node is.
fn coordinates(subject: &NodeSubject<'_>) -> (Option<String>, Option<usize>) {
    match subject {
        NodeSubject::State(state) => (Some(state.id.clone()), None),
        NodeSubject::Transition { state, index, .. } => (Some(state.id.clone()), Some(*index)),
        NodeSubject::Invoke { state, index, .. } => (Some(state.id.clone()), Some(*index)),
        NodeSubject::Action { state, site, .. } => (
            Some(state.id.clone()),
            match site {
                // Only a transition's own content carries an edge index;
                // entry and exit content belongs to the state itself.
                ActionSite::Transition { index } => Some(*index),
                ActionSite::Entry | ActionSite::Exit => None,
                // <initial> and <history> default transitions carry
                // executable content with no transition index of their
                // own; they belong to the state.
                ActionSite::Initial | ActionSite::HistoryDefault { .. } => None,
            },
        ),
        // A top-level <script> sits in no state.
        NodeSubject::GlobalScript { .. } => (None, None),
    }
}

/// The overlay for this document.
///
/// Composes the shared walk with the acceptance report's closure. There
/// is deliberately no third source of truth here — see the module header.
pub fn overlay(model: &SCXMLModel) -> AnnotationOverlay {
    let walked = walk_nodes(model);

    let mut nodes: Vec<OverlayNode> = Vec::with_capacity(walked.len());
    let mut claimed: std::collections::BTreeSet<String> = std::collections::BTreeSet::new();

    for node in &walked {
        let (state_id, transition_index) = coordinates(&node.subject);
        for id in &node.record.requirement_ids {
            claimed.insert((*id).to_string());
        }
        nodes.push(OverlayNode {
            node_path: node.record.node_path.clone(),
            node_type: node.record.node_type,
            state_id,
            transition_index,
            requirements: node
                .record
                .requirement_ids
                .iter()
                .map(|id| (*id).to_string())
                .collect(),
        });
    }

    // The closure, per requirement, from the report's own function.
    let fragments = claimed
        .into_iter()
        .map(|requirement| {
            let dependencies = fragment(model, &requirement)
                .into_iter()
                .map(|dep| OverlayDependency {
                    node_path: dep.node_path,
                    relation: relation_slug(dep.relation),
                })
                .collect();
            OverlayFragment {
                requirement,
                dependencies,
            }
        })
        .collect();

    AnnotationOverlay {
        v: OVERLAY_WIRE_VERSION,
        nodes,
        fragments,
    }
}
